File: apps/agent-macos/Sources/EgressViewAgentCore/ThreatFeedDownloader.swift
import Foundation

/// Downloads the same public threat feeds the Hub reads, for agents that have
/// no Hub.
///
/// These are plain list downloads, not lookup services: no key, no query, and
/// **no destination from this Mac is sent anywhere**. What it does reveal is
/// that this Mac asked at all, which is why it stays something the user turns
/// on rather than something done for them.
public struct ThreatFeedDownloader: Sendable {
    /// The same four feeds `src/threat-intel.js` uses. Kept identical on
    /// purpose: a standalone agent and a Hub-connected one should not disagree
    /// about what is dangerous.
    public static let feeds: [(url: URL, kind: FeedKind, source: String)] = [
        (URL(string: "https://feodotracker.abuse.ch/downloads/ipblocklist.csv")!, .feodo, "feodo"),
        (URL(string: "https://threatfox.abuse.ch/export/csv/ip-port/recent/")!, .threatfox, "threatfox"),
        (URL(string: "https://urlhaus.abuse.ch/downloads/csv_recent/")!, .urlhaus, "urlhaus"),
        (URL(string: "https://www.spamhaus.org/drop/drop.txt")!, .spamhausDrop, "spamhaus"),
    ]

    public enum FeedKind: Sendable {
        case feodo
        case threatfox
        case urlhaus
        case spamhausDrop
    }

    public enum DownloadError: Error, Equatable {
        /// Every feed failed. One failing is normal and is not reported.
        case allFeedsFailed
    }

    private let transport: any GeoCacheTransport

    public init(transport: any GeoCacheTransport = URLSessionGeoCacheTransport(timeout: 60)) {
        self.transport = transport
    }

    public func download() async throws -> [ThreatIndicator] {
        var indicators: [ThreatIndicator] = []
        var anySucceeded = false
        for feed in Self.feeds {
            var request = URLRequest(url: feed.url)
            request.httpMethod = "GET"
            request.httpShouldHandleCookies = false
            guard let (data, response) = try? await transport.fetch(request),
                  response.statusCode == 200,
                  let text = String(data: data, encoding: .utf8)
            else {
                // One feed being down is ordinary. Failing the whole refresh
                // for it would throw away three working lists.
                continue
            }
            anySucceeded = true
            indicators.append(contentsOf: Self.parse(text, kind: feed.kind, source: feed.source))
        }
        guard anySucceeded else { throw DownloadError.allFeedsFailed }
        return indicators
    }

    public static func parse(
        _ text: String, kind: FeedKind, source: String
    ) -> [ThreatIndicator] {
        switch kind {
        case .feodo: return parseFeodo(text, source: source)
        case .threatfox: return parseThreatFox(text, source: source)
        case .urlhaus: return parseURLhaus(text, source: source)
        case .spamhausDrop: return parseSpamhausDrop(text, source: source)
        }
    }

    /// `first_seen_utc,dst_ip,dst_port,…,malware`
    static func parseFeodo(_ text: String, source: String) -> [ThreatIndicator] {
        lines(text).compactMap { line in
            guard !line.hasPrefix("#") else { return nil }
            let fields = csvFields(String(line))
            guard fields.count > 1, isPlausibleIPv4(fields[1]) else { return nil }
            let malware = fields.count > 4 ? fields[4] : ""
            return ThreatIndicator(
                kind: .ip, value: fields[1], source: source,
                tag: malware.isEmpty ? "botnet C2" : "\(malware) C2"
            )
        }
    }

    /// `first_seen_utc,ioc_id,ioc_value(ip:port),ioc_type,threat_type,…,malware`
    static func parseThreatFox(_ text: String, source: String) -> [ThreatIndicator] {
        lines(text).compactMap { line in
            guard !line.hasPrefix("#") else { return nil }
            let fields = csvFields(String(line))
            guard fields.count > 2 else { return nil }
            // Strips the port: the indicator is the host, and keeping the port
            // would stop it matching the same host reached on another one.
            let value = fields[2].split(separator: ":").first.map(String.init) ?? fields[2]
            guard isPlausibleIPv4(value) else { return nil }
            let malware = fields.count > 5 ? fields[5] : ""
            return ThreatIndicator(
                kind: .ip, value: value, source: source,
                tag: malware.isEmpty ? "malware infrastructure" : malware
            )
        }
    }

    /// `id,dateadded,url,url_status,…,threat,…`
    static func parseURLhaus(_ text: String, source: String) -> [ThreatIndicator] {
        lines(text).compactMap { line in
            guard !line.hasPrefix("#") else { return nil }
            let fields = csvFields(String(line))
            guard fields.count > 2, let host = URL(string: fields[2])?.host, !host.isEmpty else {
                return nil
            }
            let isAddress = isPlausibleIPv4(host)
            return ThreatIndicator(
                kind: isAddress ? .ip : .domain,
                value: isAddress ? host : host.lowercased(),
                source: source,
                tag: "malware distribution"
            )
        }
    }

    /// `1.2.3.0/24 ; SBL123456`
    static func parseSpamhausDrop(_ text: String, source: String) -> [ThreatIndicator] {
        lines(text).compactMap { line in
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard !trimmed.isEmpty, !trimmed.hasPrefix(";"), !trimmed.hasPrefix("#") else {
                return nil
            }
            let cidr = trimmed.split(separator: ";").first?
                .trimmingCharacters(in: .whitespaces) ?? ""
            guard ThreatMatcher.parseCIDR(cidr) != nil else { return nil }
            return ThreatIndicator(
                kind: .cidr, value: cidr, source: source,
                tag: "Spamhaus DROP (hijacked network)"
            )
        }
    }

    /// Splits a feed into lines, whatever it ends them with.
    ///
    /// `split(separator: "\n")` looks like it does this and does not. Swift
    /// treats `"\r\n"` as a **single** `Character`, so a CRLF file contains no
    /// `"\n"` character at all: the whole download comes back as one line,
    /// that line starts with `#`, and every parser here drops it as a comment.
    ///
    /// Three of the four feeds ship CRLF. They returned **zero indicators from
    /// the day this was written until 2026-08-20**, with no error anywhere --
    /// a failed feed is skipped by design, and a feed that parses to nothing is
    /// indistinguishable from one that had nothing to report. Only Spamhaus,
    /// which ships LF, ever worked.
    static func lines(_ text: String) -> [Substring] {
        text.split(whereSeparator: \.isNewline)
    }

    /// Minimal RFC 4180 reader: enough for feeds that quote fields containing
    /// commas, which these do.
    static func csvFields(_ line: String) -> [String] {
        var fields: [String] = []
        var current = ""
        var inQuotes = false
        var iterator = line.makeIterator()
        while let character = iterator.next() {
            if character == "\"" {
                inQuotes.toggle()
            } else if character == "," && !inQuotes {
                fields.append(current.trimmingCharacters(in: .whitespaces))
                current = ""
            } else {
                current.append(character)
            }
        }
        fields.append(current.trimmingCharacters(in: .whitespaces))
        return fields
    }

    static func isPlausibleIPv4(_ text: String) -> Bool {
        ThreatMatcher.ipv4ToNumber(text) != nil
    }
}

File: apps/agent-windows/src/EgressView.Agent.Core/ThreatFeedDownloader.cs
using System.Net;

namespace EgressView.Agent.Core;

/// What came back from the public feeds, including which ones did not.
///
/// The count alone is not enough to judge by. On the Mac, three of the four
/// feeds returned nothing for the whole life of that code and the screen went
/// on showing a total, because a feed that yields no rows looks exactly like a
/// feed with nothing to report. Whoever turned it on could not tell.
public sealed record ThreatFeedResult(IReadOnlyList<ThreatIndicator> Indicators, IReadOnlyList<string> MissingSources)
{
    public bool IsComplete => MissingSources.Count == 0;
}

public sealed class ThreatFeedException(string message) : Exception(message);

/// Downloads the same public threat feeds the Hub reads, for agents that have
/// no Hub.
///
/// These are plain list downloads, not lookup services: no key, no query, and
/// <b>no destination from this PC is sent anywhere</b>. What it does reveal is
/// that this PC asked at all, which is why it stays something the person turns
/// on rather than something done for them.
///
/// Ported from the Mac Agent, feeds, parsers and judgements alike. A
/// standalone agent and a Hub-connected one must not disagree about what is
/// dangerous, and neither must the two platforms.
public sealed class ThreatFeedDownloader(HttpClient? http = null)
{
    /// The same four feeds the Hub's threat-intel reads.
    public static readonly (string Url, string Kind, string Source)[] Feeds =
    [
        ("https://feodotracker.abuse.ch/downloads/ipblocklist.csv", "feodo", "feodo"),
        ("https://threatfox.abuse.ch/export/csv/ip-port/recent/", "threatfox", "threatfox"),
        ("https://urlhaus.abuse.ch/downloads/csv_recent/", "urlhaus", "urlhaus"),
        ("https://www.spamhaus.org/drop/drop.txt", "spamhausDrop", "spamhaus"),
    ];

    /// Feeds that legitimately publish nothing for long stretches, so an empty
    /// result from them is a fact rather than a fault.
    ///
    /// Feodo Tracker's blocklist has been empty since 2026-03-04. Treating
    /// that as a broken feed would put a warning on screen that never clears,
    /// and a warning that never clears is one nobody reads.
    public static readonly IReadOnlySet<string> PublishesEmptyLists =
        new HashSet<string>(StringComparer.Ordinal) { "feodo" };

    /// Services that host other people's files.
    ///
    /// URLhaus lists the URL malware was served from. A match on one of these
    /// says something was hosted there, not that the host is hostile, so it is
    /// low confidence. The same list exists on the Hub and on the Mac.
    public static readonly IReadOnlySet<string> LowConfidenceDomains =
        new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "github.com", "raw.githubusercontent.com", "gist.githubusercontent.com",
            "gitlab.com", "bitbucket.org",
            "drive.google.com", "docs.google.com", "storage.googleapis.com",
            "dropbox.com", "dl.dropboxusercontent.com",
            "onedrive.live.com", "1drv.ms",
            "cdn.discordapp.com", "media.discordapp.net",
            "pastebin.com", "paste.ee",
            "transfer.sh", "anonfiles.com",
            "amazonaws.com", "s3.amazonaws.com",
            "cloudfront.net", "azureedge.net", "blob.core.windows.net",
            "archive.org",
        };

    private readonly HttpClient http = http ?? CreateClient();

    private static HttpClient CreateClient()
    {
        var client = new HttpClient(new SocketsHttpHandler
        {
            AllowAutoRedirect = true,
            AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate,
        })
        { Timeout = TimeSpan.FromMinutes(1) };
        AgentEnrollmentClient.EnsureUserAgent(client);
        return client;
    }

    public async Task<ThreatFeedResult> DownloadAsync(CancellationToken cancellationToken = default)
    {
        var indicators = new List<ThreatIndicator>();
        var missing = new List<string>();
        var anySucceeded = false;
        foreach (var (url, kind, source) in Feeds)
        {
            string text;
            try
            {
                using var response = await http.GetAsync(url, cancellationToken);
                if (response.StatusCode != HttpStatusCode.OK) { missing.Add(source); continue; }
                text = await response.Content.ReadAsStringAsync(cancellationToken);
            }
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested) { throw; }
            catch (Exception)
            {
                // One feed being down is ordinary, so the refresh continues --
                // but it is recorded rather than shrugged off.
                missing.Add(source);
                continue;
            }
            anySucceeded = true;
            var parsed = Parse(text, kind, source);
            if (parsed.Count == 0 && !PublishesEmptyLists.Contains(source))
            {
                // Downloaded and understood nothing. That is a parser or a
                // changed format, not a quiet day, and it is the exact failure
                // that went unnoticed for the life of the Mac's version.
                missing.Add(source);
                continue;
            }
            indicators.AddRange(parsed);
        }
        if (!anySucceeded) throw new ThreatFeedException("Every public threat feed failed.");
        return new ThreatFeedResult(indicators, missing);
    }

    /// Finds a column by its name in the feed's own header line.
    ///
    /// The Mac reads the malware family from a fixed index and takes the wrong
    /// column: for Feodo's `first_seen_utc,dst_ip,dst_port,c2_status,
    /// last_online,malware` it reads `last_online`, so its indicators are
    /// tagged "2026-08-01 C2". Nothing caught it because its tests assert the
    /// address and the kind and never the tag.
    ///
    /// Reading the header instead of counting commas is both correct today and
    /// unbroken by a feed that adds a column tomorrow.
    private static int ColumnIndex(string[] lines, string name)
    {
        foreach (var line in lines)
        {
            var header = line.TrimStart('#', ' ');
            var fields = CsvFields(header);
            for (var index = 0; index < fields.Count; index++)
                if (string.Equals(fields[index], name, StringComparison.OrdinalIgnoreCase)) return index;
            // Only the first non-empty line is a candidate header.
            if (header.Length > 0 && !line.TrimStart().StartsWith('#')) break;
        }
        return -1;
    }

    private static string FieldOrEmpty(List<string> fields, int index) =>
        index >= 0 && index < fields.Count ? fields[index] : string.Empty;

    public static IReadOnlyList<ThreatIndicator> Parse(string text, string kind, string source) => kind switch
    {
        "feodo" => ParseFeodo(text, source),
        "threatfox" => ParseThreatFox(text, source),
        "urlhaus" => ParseUrlhaus(text, source),
        "spamhausDrop" => ParseSpamhausDrop(text, source),
        _ => [],
    };

    private static List<ThreatIndicator> ParseFeodo(string text, string source)
    {
        var lines = Lines(text);
        var addressColumn = ColumnIndex(lines, "dst_ip");
        var malwareColumn = ColumnIndex(lines, "malware");
        if (addressColumn < 0) addressColumn = 1;
        var result = new List<ThreatIndicator>();
        foreach (var line in lines)
        {
            if (line.StartsWith('#')) continue;
            var fields = CsvFields(line);
            var address = FieldOrEmpty(fields, addressColumn);
            if (!IsPlausibleIpv4(address)) continue;
            var malware = FieldOrEmpty(fields, malwareColumn);
            result.Add(new ThreatIndicator("ip", address, source,
                malware.Length == 0 ? "botnet C2" : malware + " C2", "high"));
        }
        return result;
    }

    private static List<ThreatIndicator> ParseThreatFox(string text, string source)
    {
        var lines = Lines(text);
        var valueColumn = ColumnIndex(lines, "ioc_value");
        var malwareColumn = ColumnIndex(lines, "malware_printable");
        if (malwareColumn < 0) malwareColumn = ColumnIndex(lines, "malware");
        if (valueColumn < 0) valueColumn = 2;
        var result = new List<ThreatIndicator>();
        foreach (var line in lines)
        {
            if (line.StartsWith('#')) continue;
            var fields = CsvFields(line);
            // Strips the port: the indicator is the host, and keeping the port
            // would stop it matching the same host reached on another one.
            var value = FieldOrEmpty(fields, valueColumn).Split(':')[0];
            if (!IsPlausibleIpv4(value)) continue;
            var malware = FieldOrEmpty(fields, malwareColumn);
            result.Add(new ThreatIndicator("ip", value, source,
                malware.Length == 0 ? "malware infrastructure" : malware, "high"));
        }
        return result;
    }

    private static List<ThreatIndicator> ParseUrlhaus(string text, string source)
    {
        var lines = Lines(text);
        var urlColumn = ColumnIndex(lines, "url");
        if (urlColumn < 0) urlColumn = 2;
        var result = new List<ThreatIndicator>();
        foreach (var line in lines)
        {
            if (line.StartsWith('#')) continue;
            var fields = CsvFields(line);
            var raw = FieldOrEmpty(fields, urlColumn);
            if (raw.Length == 0 || !Uri.TryCreate(raw, UriKind.Absolute, out var uri)) continue;
            var host = uri.Host;
            if (host.Length == 0) continue;
            var isAddress = IsPlausibleIpv4(host);
            var confidence = isAddress ? "high" : ConfidenceForHost(host);
            result.Add(new ThreatIndicator(isAddress ? "ip" : "domain",
                isAddress ? host : host.ToLowerInvariant(), source,
                confidence == "low" ? "file hosting service (low confidence)" : "malware distribution",
                confidence));
        }
        return result;
    }

    private static List<ThreatIndicator> ParseSpamhausDrop(string text, string source)
    {
        var result = new List<ThreatIndicator>();
        foreach (var line in Lines(text))
        {
            var trimmed = line.Trim();
            if (trimmed.Length == 0 || trimmed.StartsWith(';') || trimmed.StartsWith('#')) continue;
            var cidr = trimmed.Split(';')[0].Trim();
            if (!IsPlausibleCidr(cidr)) continue;
            result.Add(new ThreatIndicator("cidr", cidr, source, "Spamhaus DROP (hijacked network)", "high"));
        }
        return result;
    }

    /// Matches the Hub and the Mac: the host itself, or its last two labels.
    public static string ConfidenceForHost(string host)
    {
        var lowered = host.ToLowerInvariant();
        var labels = lowered.Split('.');
        var parent = labels.Length >= 2 ? string.Join('.', labels[^2..]) : lowered;
        return LowConfidenceDomains.Contains(lowered) || LowConfidenceDomains.Contains(parent) ? "low" : "high";
    }

    /// Splits a feed into lines, whatever it ends them with.
    ///
    /// Three of the four feeds ship CRLF. On the Mac they returned zero
    /// indicators from the day that code was written until 2026-08-20, with no
    /// error anywhere: the whole download came back as one line, that line
    /// started with a comment marker, and every parser dropped it. Only
    /// Spamhaus, which ships LF, ever worked.
    public static string[] Lines(string text) =>
        text.Split(['\r', '\n'], StringSplitOptions.RemoveEmptyEntries);

    /// Minimal RFC 4180 reader: enough for feeds that quote fields containing
    /// commas, which these do.
    public static List<string> CsvFields(string line)
    {
        var fields = new List<string>();
        var current = new System.Text.StringBuilder();
        var inQuotes = false;
        foreach (var character in line)
        {
            if (character == '"') inQuotes = !inQuotes;
            else if (character == ',' && !inQuotes) { fields.Add(current.ToString().Trim()); current.Clear(); }
            else current.Append(character);
        }
        fields.Add(current.ToString().Trim());
        return fields;
    }

    private static bool IsPlausibleIpv4(string text) =>
        IPAddress.TryParse(text, out var address) &&
        address.AddressFamily == System.Net.Sockets.AddressFamily.InterNetwork;

    private static bool IsPlausibleCidr(string text)
    {
        var parts = text.Split('/');
        return parts.Length == 2 && int.TryParse(parts[1], out var prefix) && prefix is >= 0 and <= 32
            && IsPlausibleIpv4(parts[0]);
    }
}
